### PR TITLE
added fix to prevent None values being used in Default variables

### PR DESCRIPTION
**Pull request by sagagliardo**

### CODE REVIEW

**Comment by sagagliardo** [comment]
None values can make their way into the DEFAULT_HOSTNAME and DEFAULT_PORT variables which are set in core.py. These variables will not default to localhost and 8020 if a value is set (in this case None) at all.

Before this I've just added in a requirement that they not be None from the urlparse, to prevent these dictionary values from even being set.

**Comment by martindurant** [comment]
Thank you for the contribution! It would be nice to see a test that demonstrates the previous bug.
Do you think that `hdfs_conf()`should additionally take a parameter for where to look for configuration, rather than assuming the environment variables are correctly set?

**Comment by sagagliardo** [comment]
Hey Matt. Here is an example of what happens:

```
admin:hdfs3$ ls
__init__.py      compatibility.py core-site.xml    core.py          hdfs-site.xml    lib.py           mapping.py       tests            utils.py
admin:hdfs3$ python core.py
Traceback (most recent call last):
  File "core.py", line 58, in <module>
    DEFAULT_PORT = int(conf.get('port', 8020))
TypeError: int() argument must be a string or a number, not 'NoneType'
```

When you have a HA setup, like in this particular example, a port isn't listed which causes the above problem.

from core-site.xml
```
    <name>fs.defaultFS</name>
    <value>hdfs://hanameservice1</value>
```

Additionally, to answer your question, that may be a good idea for a separate PR. That is, rather than assume '/hadoop/conf', we could have in the future hdfs_conf() take a single optional variable with something like hdfs_conf(conf_dir='/hadoop/conf') so it can be overwritten if desired.

**Comment by martindurant** [comment]
Like this?

```
def test_no_conf():
    os.environ.pop('HADOOP_CONF_DIR', '')
    os.environ.pop('HADOOP_INSTALL', '')
    conf = hdfs3.core.conf()
    assert conf['host'] is not None
    assert conf['port'] is not None
```

**Comment by sagagliardo** [comment]
Oh, sorry. yeah forgot the test part.

I think it would be better to do this, to be honest:

```
def test_conf_defaults():
    os.environ.pop('HADOOP_CONF_DIR', '')
    os.environ.pop('HADOOP_INSTALL', '')
    conf = hdfs3.core.conf()
    assert DEFAULT_HOSTNAME is not None
    assert DEFAULT_PORT is not None
```

Because that makes sure the defaults are being set in the absence of those values.

**Comment by martindurant** [comment]
The way it stands now, `hdfs3.core.conf` is justa  dictionary, not a function, and DEFAULT_HOSTNAME, DEFAULT_PORT are only set at import time, so changing the env later will not affect them. Of course, they *could* be globals set from within `hdfs_conf()`.

**Comment by sagagliardo** [comment]
Ah nvm you're right. Yeah you're calling from outside of it. 

I concur that your original solution would be a good test for this then. My mistake. This is about as good we're going to get, what you specified, per below:

```
def test_no_conf():
    os.environ.pop('HADOOP_CONF_DIR', '')
    os.environ.pop('HADOOP_INSTALL', '')
    conf = hdfs3.core.conf()
    assert conf['host'] is not None
    assert conf['port'] is not None
```

**Comment by martindurant** [comment]
OK, then please add that, and I'll merge later.
I may get around to doing the more thorough version where you pass an optional conf directory.

**Comment by sagagliardo** [comment]
Done, added.

**Comment by martindurant** [comment]
I would have conf_dir=None, and in the function

```
if conf_dir is None:
    confd = os.get...
else:
    confd = conf_dir
```

Otherwise the environment variable takes precedence over the passed argument, which is probably the wrong way around. 

This could be tested too by making a temporary directory and putting an xml extract in it with known value for the port. Looking at the code, does this even work at all? I can't find hdfs|core-site.xml example files with quite the structure that hdfs_conf() seems to want.

**Comment by sagagliardo** [comment]
Also good point. I guess I'm assuming they'll always want to use the environment variables and just have a different extension. Will switch to that.

File: hdfs3/tests/test_hdfs3.py [hunk 1]
```diff
@@ -54,6 +54,12 @@ def test_simple(hdfs):
         assert len(data) == len(out)
         assert out == data
 
+def test_no_conf():
+    os.environ.pop('HADOOP_CONF_DIR', '')
+    os.environ.pop('HADOOP_INSTALL', '')
+    conf = hdfs3.core.conf()
```
Comment (martindurant) [inline hdfs3/tests/test_hdfs3.py]: Sorry, my fault: something needs to be added to the imports at the head of the file.

**Comment by martindurant** [comment]
With that import you need 

    conf = hdfs3.core.conf()
    - >
    conf = conf()

(which masks the name, but that's probably OK)

**Comment by martindurant** [comment]
Hah, doesn't work: conf is no longer a local variable, so it doesn't pick up the global module :)
Perhaps you can run pytest locally to make sure it's working?

**Comment by sagagliardo** [comment]
I put some logic in the test so if conf['port'] and hostname aren't determined from the config (IE, unset, which is fine since defaults will take over) to just ignore the testing for None values. But if they are set, to ensure they're not None. Hopefully that should resolve the test failures in this case.

**Comment by martindurant** [comment]
OK, that will do, thank you!

**Comment by sagagliardo** [comment]
Awesome. Thanks!

**Pull request merged**